src/components/Personalization/helper/index.js

```diff
@@ -24,22 +26,32 @@ export { default as setAttributes } from "./setAttributes";
 export { default as swapImage } from "./swapImage";
 export { default as rearrangeChildren } from "./rearrangeChildren";
 
-export const createAction = (collect, renderFunc) => {
+const renderContent = (elements, content, renderFunc) => {
+  const executions = elements.map(element =>
+    Promise.resolve(renderFunc(element, content))
+  );
+
+  return Promise.all(executions);
+};
+
+export const createAction = renderFunc => {
   return (settings, event) => {
-    const { elements, prehidingSelector } = event;
-    const { content, meta } = settings;
-    const executions = elements.map(element =>
-      Promise.resolve(renderFunc(element, content))
-    );
+    const { notify } = event;
+    const { selector, prehidingSelector, content } = settings;
+
+    hideElements(prehidingSelector);
 
-    return Promise.all(executions)
+    return awaitSelector(selector, selectNodesWithEq)
+      .then(elements => renderContent(elements, content, renderFunc))
       .then(() => {
-        // Success, unhide elements and notify
+        // if everything is OK, notify and show elements
+        notify();
         showElements(prehidingSelector);
-        collect(meta);
       })
       .catch(() => {
-        // Something went horribly wrong, unhide elements
+        // in case of awaiting timing or error, we need to remove the style tag
+        // hence showing the pre-hidden elements and notify
+        notify();
```

Comment:
if showElements throws an error, notify could be called twice for one Action.  Additionally, the `then` and `catch` are identical.  Can we combine them into a finally?
```
return awaitSelector(selector, selectNodesWithEq)
      .then(elements => renderContent(elements, content, renderFunc))
      .finally(() => {
        notify();
        showElements(prehidingSelector);
      });
```

Reply:
Regarding `finally`, I made a decision a while back to not use `finally` elsewhere in the code: https://github.com/adobe/alloy/blob/67a7e1c125a37986814923d591fadf4308794c2b/src/core/network/createNetwork.js#L122-L124

Like my comment says, I made that decision because it's a relatively new feature and I wasn't sure if polyfills that customers may be using supported it. I have no idea if I made the right decision though and am totally cool with going either way on it. If you update this one to use `finally` though, please update the code I linked to as well.